Python/bytecodes.c

```diff
@@ -279,73 +279,111 @@ dummy_func(
 
         family(binary_op, INLINE_CACHE_ENTRIES_BINARY_OP) = {
             BINARY_OP,
-            BINARY_OP_ADD_FLOAT,
+            BINARY_OP_MULTIPLY_INT,
             BINARY_OP_ADD_INT,
-            BINARY_OP_ADD_UNICODE,
-            // BINARY_OP_INPLACE_ADD_UNICODE,  // This is an odd duck.
+            BINARY_OP_SUBTRACT_INT,
             BINARY_OP_MULTIPLY_FLOAT,
-            BINARY_OP_MULTIPLY_INT,
+            BINARY_OP_ADD_FLOAT,
             BINARY_OP_SUBTRACT_FLOAT,
-            BINARY_OP_SUBTRACT_INT,
+            BINARY_OP_ADD_UNICODE,
+            // BINARY_OP_INPLACE_ADD_UNICODE,  // See comments at that opcode.
         };
 
-
-        inst(BINARY_OP_MULTIPLY_INT, (unused/1, left, right -- prod)) {
+        op(_BINARY_OP_INT_GUARD, (left, right -- left, right)) {
             DEOPT_IF(!PyLong_CheckExact(left), BINARY_OP);
             DEOPT_IF(!PyLong_CheckExact(right), BINARY_OP);
+        }
```

Comment:
It seems to me that there might be value in splitting these types of checks up: one for guarding the top of stack, and one for guarding the second item on the stack.

That way we could successfully remove uops for cases like `1 + rhs` or `lhs += 1`, where they type of one argument is known (or inferred).

Reply:
Just a heads up: we plan to just have a generic `CHECK_INT x` that checks the `x`th stack entry is indeed an int. The main reason is that it would save us an opcode.

Reply:
That makes sense. Just need space for a couple extra opargs. I wonder if it would make sense to have "hardcoded" or "shared" opargs as part of macro instructions.

So we could have:

```
macro(BINARY_OP) = _BINARY_OP_SPECIALIZE(oparg) + _BINARY_OP_ACTION(oparg);
macro(BINARY_OP_ADD_INT) = _CHECK_INT(1) + _CHECK_INT(2) + _BINARY_OP_ADD_INT(oparg);
```

Reply:
@gvanrossum, thoughts on something like the above?

Reply:
I forgot to also mention that we encountered many situations where one type is known but the other isnt. So it helps to have some granularity.

Reply:
Okay, I think I'll follow Mark's guidance, and keep the current structure.